core/embed/util/fwutils/fwutils.c

```diff
@@ -21,61 +21,88 @@
 #include "flash_area.h"
 
 #ifdef KERNEL_MODE
 
 #define FW_HASHING_CHUNK_SIZE 1024
 
-secbool firmware_calc_hash(const uint8_t* challenge, size_t challenge_len,
-                           uint8_t* hash, size_t hash_len,
-                           firmware_hash_callback_t callback,
-                           void* callback_context) {
-  BLAKE2S_CTX ctx;
+typedef struct {
+  bool initialized;
+  BLAKE2S_CTX blake;
+  uint32_t fw_offset;
+  uint32_t fw_size;
+
+} firmware_hash_context_t;
+
+static firmware_hash_context_t g_hash_context[SYSTASK_MAX_TASKS];
+
+int firmware_hash_start(const uint8_t* challenge, size_t challenge_len) {
+  firmware_hash_context_t* ctx = &g_hash_context[systask_id(systask_active())];
+
+  int err;
 
   if (challenge_len != 0) {
-    if (blake2s_InitKey(&ctx, BLAKE2S_DIGEST_LENGTH, challenge,
-                        challenge_len) != 0) {
-      return secfalse;
-    }
+    err = blake2s_InitKey(&ctx->blake, BLAKE2S_DIGEST_LENGTH, challenge,
+                          challenge_len);
   } else {
-    blake2s_Init(&ctx, BLAKE2S_DIGEST_LENGTH);
+    err = blake2s_Init(&ctx->blake, BLAKE2S_DIGEST_LENGTH);
+  }
+
+  if (err != 0) {
+    return -1;
   }
 
-  uint32_t firmware_size = flash_area_get_size(&FIRMWARE_AREA);
-  uint32_t chunks = firmware_size / FW_HASHING_CHUNK_SIZE;
+  ctx->fw_offset = 0;
+  ctx->fw_size = flash_area_get_size(&FIRMWARE_AREA);
 
-  ensure((firmware_size % FW_HASHING_CHUNK_SIZE == 0) * sectrue,
+  ensure((ctx->fw_size % FW_HASHING_CHUNK_SIZE == 0) * sectrue,
          "Cannot compute FW hash.");
 
-  for (int i = 0; i < chunks; i++) {
-    if (callback != NULL && (i % 128 == 0)) {
-      callback(callback_context, i, chunks);
-    }
+  ctx->initialized = true;
+  return 0;
+}
 
-    const void* data = flash_area_get_address(
-        &FIRMWARE_AREA, i * FW_HASHING_CHUNK_SIZE, FW_HASHING_CHUNK_SIZE);
+int firmware_hash_continue(uint8_t* hash, size_t hash_len) {
+  firmware_hash_context_t* ctx = &g_hash_context[systask_id(systask_active())];
 
-    if (data == NULL) {
-      return secfalse;
-    }
+  memset(hash, 0, hash_len);
 
-    blake2s_Update(&ctx, data, FW_HASHING_CHUNK_SIZE);
+  if (!ctx->initialized || ctx->fw_offset >= ctx->fw_size) {
+    return -1;
   }
 
-  if (callback != NULL) {
-    callback(callback_context, chunks, chunks);
+  int n_chunks = 128;
+
+  while (ctx->fw_offset < ctx->fw_size && n_chunks > 0) {
+    const void* chunk_ptr = flash_area_get_address(
+        &FIRMWARE_AREA, ctx->fw_offset, FW_HASHING_CHUNK_SIZE);
+
+    int err = blake2s_Update(&ctx->blake, chunk_ptr, FW_HASHING_CHUNK_SIZE);
+    if (err != 0) {
+      ctx->initialized = false;
+      return -1;
+    }
+
+    ctx->fw_offset += FW_HASHING_CHUNK_SIZE;
+    --n_chunks;
   }
 
-  if (blake2s_Final(&ctx, hash, hash_len) != 0) {
-    return secfalse;
+  if (ctx->fw_offset >= ctx->fw_size) {
+    int err = blake2s_Final(&ctx->blake, hash, hash_len);
+    if (err != 0) {
+      ctx->initialized = false;
```

Comment:
should we maybe se initialized to false even on successful calculation of final hash? to indicate that we are finished